promslog/slog_test.go

```diff
@@ -0,0 +1,152 @@
+	}
+}
+
+func TestUnmarshallBadLevel(t *testing.T) {
+	l := &AllowedLevel{}
+	err := yaml.Unmarshal([]byte(`debugg`), l)
+	if err == nil {
+		t.Error("expected error")
+	}
+	expErr := `unrecognized log level "debugg"`
+	if err.Error() != expErr {
+		t.Errorf("expected error %s, got %s", expErr, err.Error())
+	}
+	if l.s != "" {
+		t.Errorf("expected empty level, got %s", l.s)
+	}
+}
+
+func getLogEntryLevelCounts(s string, re *regexp.Regexp) map[string]int {
+	counters := make(map[string]int)
+	lines := strings.Split(s, "\n")
+	for _, line := range lines {
+		matches := re.FindStringSubmatch(line)
+		if len(matches) > 1 {
+			levelIndex := re.SubexpIndex("LevelValue")
+
+			counters[strings.ToLower(matches[levelIndex])]++
+		}
+	}
+
+	return counters
+}
+
+func TestDynamigLevels(t *testing.T) {
+	var buf bytes.Buffer
+	wantedLevelCounts := map[string]int{"info": 1, "debug": 1}
+
+	tests := map[string]struct {
+		logStyle         string
+		logStyleRegexp   *regexp.Regexp
+		wantedLevelCount map[string]int
+	}{
+		"empty_log_style":  {logStyle: "", logStyleRegexp: slogStyleLogRegexp, wantedLevelCount: wantedLevelCounts},
+		"bad_log_style":    {logStyle: "foo", logStyleRegexp: slogStyleLogRegexp, wantedLevelCount: wantedLevelCounts},
+		"slog_log_style":   {logStyle: "slog", logStyleRegexp: slogStyleLogRegexp, wantedLevelCount: wantedLevelCounts},
+		"go-kit_log_style": {logStyle: "go-kit", logStyleRegexp: goKitStyleLogRegexp, wantedLevelCount: wantedLevelCounts},
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			buf.Reset() // Ensure buf is reset prior to tests
+			config := &Config{ioWriter: &buf, Style: &AllowedStyle{s: tc.logStyle}}
+			logger := New(config)
+
+			// Test that log level can be adjusted on-the-fly to debug and that a
+			// log entry can be written to the file.
+			if err := config.Level.Set("debug"); err != nil {
+				t.Fatal(err)
+			}
+			logger.Info("info", "hello", "world")
+			logger.Debug("debug", "hello", "world")
+
+			counts := getLogEntryLevelCounts(buf.String(), tc.logStyleRegexp)
+			require.Equal(t, tc.wantedLevelCount["info"], counts["info"], "info log successfully detected")
+			require.Equal(t, tc.wantedLevelCount["debug"], counts["debug"], "debug log successfully detected")
+			// Print logs for humans to see, if needed.
+			fmt.Println(buf.String())
+			buf.Reset()
+
+			// Test that log level can be adjusted on-the-fly to info and that a
+			// subsequent call to write a debug level log is _not_ written to the
+			// file.
+			if err := config.Level.Set("info"); err != nil {
+				t.Fatal(err)
+			}
+			logger.Info("info", "hello", "world")
+			logger.Debug("debug", "hello", "world")
+
+			counts = getLogEntryLevelCounts(buf.String(), tc.logStyleRegexp)
+			require.Equal(t, tc.wantedLevelCount["info"], counts["info"], "info log successfully detected")
+			require.NotEqual(t, tc.wantedLevelCount["debug"], counts["debug"], "extra debug log detected")
```

Comment:
```suggestion
			require.NotEqual(t, 0, counts["debug"], "extra debug log detected")
```

Not sure if I missed something, but shouldn't it be 0 now that debug was disabled?

Reply:
The value should indeed be zero -- for sake of reusing the simple predefined map of values, I inverted the logical test for this final case to test `log count != 1`

Reply:
Oh, I didn't realize the Equal/NotEqual change 😛. LGTM